### DISCUSSION TITLE
How to send a follow-up email that is grouped like a conversation ?

### DISCUSSION THREAD

**Discussion by rick2ricks** [anymail/django-anymail#264] [Q&A] [answered]
Hi, 

I would like to achieve the same effect as when I reply an email using gmail or any other provider, the email is grouped into a conversation. Is it possible to do it with any anymail providers? 

ps: I want to reply the same email I have sent from anymail, not an email I have received.

Thanks.

**Comment by medmunds** [comment, marked as answer]
This isn't really specific to Anymail or any particular ESP. Each email client app has its own rules about when it groups messages into conversations. (So you might find better info searching for "grouping conversations in Gmail" in a more general forum like Stack Overflow.)

My memory (which could be wrong!) is that Gmail conversations are based almost entirely on matching the email's subject and the participants (from and to addresses). So if your first message has the subject "Your report", later messages will be grouped in the same conversation as long as you send them with the same from and to addresses and a similar subject—e.g., "Re: Your report".

Some other email clients try to match the *In-Reply-To* and/or *References* email headers. If those headers match the *Message-ID* header from an earlier email, the client will group those messages into a conversation. 

If you were replying to a message you received, you would use the value of that message's *Message-ID* for your *In-Reply-To*. But since you want to send the first email yourself, how to get the original *Message-ID* depends on the ESP.

Some ESPs will let you set your own *Message-ID* when you send the message, which makes it easy:

```python
# For first message:
first_message_id = email.utils.make_msgid()
first_message.headers["Message-ID"] = first_message_id
first_message.send()

# For later "replies":
second_message.headers["In-Reply-To"] = first_message_id
second_message.send()
```

But other ESPs insist on generating their own message ids as you send the message, and will either error or silently override your attempt to provide your own *Message-ID* header. (Unfortunately, I don't know offhand which ESPs do this.)

I believe that Mailgun and Sendinblue return their *Message-ID* values from the send call, so for those ESPs you could check `first_message.anymail_status.message_id` after the `send()`. (I think most other ESPs use opaque internal database ids for `anymail_status.message_id`, so those won't work with *In-Reply-To* headers.)

Bottom line is, you'll probably need to experiment a little with your choice of ESP and the email clients you care most about.

**Reply by rick2ricks** [reply]
Thank you for the information, I still wasn't able to accomplish it by only changing headers, but I might try a few more times.

I am using postmark by the way.

**Reply by medmunds** [reply]
Postmark's docs say they [allow setting a custom *Message-ID* header](https://postmarkapp.com/developer/user-guide/send-email-with-api#:~:text=To%20send%20an%20email%20with%20a%20custom%20Message%2DID%20header%2C%20you%20can%20pass%20the%20value%20you%20desire), so that approach should work for email clients that thread based on Message-ID. 

You may also need to set the *References* header. It's similar to *In-Reply-To*, but contains the message-id of all previous messages in the thread (so gets a little more complicated after the second message). The Wikipedia article on [*Conversation Threading*](https://en.wikipedia.org/wiki/Conversation_threading#Mechanism) quotes and links to the relevant spec.

Also, apparently Gmail (at least for GSuite/Workspace users) started requiring an accurate *References* header a few years ago. See [*How does Gmail decide to group emails into conversations*](https://support.cloudhq.net/how-does-gmail-decide-to-group-emails-into-conversations/) (not from Google) and [*Threading changes in Gmail conversation view*](http://workspaceupdates.googleblog.com/2019/03/threading-changes-in-gmail-conversation-view.html) (from Google).

And no matter what other headers say, the email subjects must be similar. You can add "Re:" or "Fwd:" (or similar) prefixes, but if you change anything else in the subject, Gmail will break it into a new conversation thread—even if  *References* or *In-Reply-To* match an earlier *Message-ID*.

If it's still not working, please show the relevant headers of two messages that you expected to be grouped but aren't (and which email client you're using), and I might be able to offer other suggestions.

**Reply by rick2ricks** [reply]
Thanks, this really helped.

Also, part of the problem is that I was testing with the same e-mail in "to" and "from".
